keps/sig-storage/5381-mutable-pv-affinity/README.md

```diff
@@ -0,0 +1,1013 @@
+I need a way to tell scheduler that the volume is now accessible in every zone,
+so that the pod can be scheduled to another zone when the current zone is down.
+
+In this case, the old affinity would be:
+```yaml
+required:
+  nodeSelectorTerms:
+  - matchExpressions:
+    - key: topology.kubernetes.io/zone
+      operator: In
+      values:
+      - cn-beijing-g
+```
+We would like to change it to:
+```yaml
+required:
+nodeSelectorTerms:
+- matchExpressions:
+  - key: topology.kubernetes.io/region
+    operator: In
+    values:
+    - cn-beijing
+```
+manually currently, hopefully integrated into CSI in the future.
+
+#### Story 2
+
+As a cluster operator, I'm conducting an upgrade to new storage category provided by our storage provider.
+However, once upgraded, the volume cannot be attached to some legacy nodes in the cluster.
+I need a way to convey this new requirement to the scheduler,
+so that my pod will not getting stuck in a `ContainerCreating` state.
+
+In this case, the old affinity would be:
+```yaml
+required:
+  nodeSelectorTerms:
+  - matchExpressions:
+    - key: provider.com/disktype.cloud_ssd
+      operator: In
+      values:
+      - available
+```
+
+Type A node only supports cloud_ssd, while Type B node supports both cloud_ssd and cloud_essd.
+We will only allow the modification if the volume is attached to type B nodes.
+And I want to make sure the Pods using new cloud_essd volume not to be scheduled to type A nodes.
+
+We would like to change the affinity to:
+```yaml
+required:
+  nodeSelectorTerms:
+  - matchExpressions:
+    - key: provider.com/disktype.cloud_essd
+      operator: In
+      values:
+      - available
+```
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+It is storage provider's responsibility to ensure that the running workload is not interrupted while the data is being moved.
+
+Whoever modifies the `PersistentVolume.spec.nodeAffinity` field should ensure that
+no running Pods on nodes with incompatible labels are using the PV.
+Kubernetes will not verify this. It is expensive and racy.
+
+If the incompatibility does happen, we don't guarantee that those Pods will continue to run without any issue.
+However, we try our best not to interrupt them:
+- For volumes that not yet present in the Node.status.volumesAttached field,
+  we fail the Pods that use them, since we are sure the Pods have never been running.
+  (see [Handling race condition](#handling-race-condition) below)
+- For CSI drivers with `requiresRepublish` set to true, we will stop calling NodePublishVolume periodically. and an event is emitted.
+- For CSI drivers with `requiresRepublish` set to false, an event is emitted on kubelet restart. Otherwise the pod should continue to run.
```

Comment:
I think we should avoid emitting the event error as it doesn't cause pod failures, but it may block volume-related node operations like filesystem expansion. Alerting the user and guiding them to fix PV affinity is necessary."